modules/local/build_sam_header.nf

```diff
@@ -0,0 +1,96 @@
+// Module that parses an NCBI assembly and assembly report and outputs
+// a SAM header template
+process BUILD_SAM_HEADER {
+    tag "${meta.id}"
+    label 'process_single'
+
+    conda "conda-forge::gawk=5.1.0"
+    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/gawk:5.1.0' :
+        'biocontainers/gawk:5.1.0'}"
+
+    input:
+    tuple val(meta), path(dict), path(report), path(source)
+
+    output:
+    tuple val(meta), path(filename_header), emit: header
+    path "versions.yml", emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    filename_header = "${prefix}.header.sam"
+
+    // Use the supplied speciesRegex or default if not provided
+    def speciesRegex = task.ext.speciesRegex ?: '# Organism name:\s*([^\\(]*)\s*\\(.*\\)'
+
+    """
+    sourcePath=\$(cat $source | tr -d '\\n')
+    genBankAccession=\$(awk '/^# GenBank assembly accession:/ { gsub("\\r", ""); print \$NF }' $report)
+    awk -v species_regex='$speciesRegex' -v genBankAccession=\$genBankAccession -v sourcePath=\$sourcePath '
+    BEGIN {
+        OFS = "\\t";
+        IFS = "\\t";
+        AS = "AS:" genBankAccession;
+        species_name = "";
+    }
+    NR == FNR {
+        if (\$0 ~ /^# Organism name:/) {
+            match(\$0, species_regex, arr);
+            species_name = arr[1];
+        }
+        if (\$0 !~ /^#/) {
+            split(\$0, fields, "\\t");
+            lookup[fields[5]] = fields[3];
+        }
+        next;
+    }
+    /^@HD/ {
+        print;
+        next;
+    }
+    /^@SQ/ {
+        split(\$0, fields, "\\t");
+        sn = "";
+        for (i in fields) {
+            if (fields[i] ~ /^SN:/) {
+                split(fields[i], sn_field, ":");
+                sn = sn_field[2];
+            }
+            if (fields[i] ~ /^UR:/) {
+                fields[i] = "UR:" sourcePath;
+            }
+        }
+        if (sn in lookup) {
+            new_field = "AN:" lookup[sn];
+        } else {
+            new_field = "AN:na";
```

Comment:
Just spotting this as I ran the "full test". Is that a standard thing to do ? Why not just skipping `AN` if there is no alias ? May `an` be considered the actual alias ??

Reply:
I wanted to make sure the file would be compatible with any downstream processing that assumes all entries have the same fields, so there would be at least a placeholder to show that no alternate name exists. I used "na" because this is what is already used as the placeholder in the NCBI `*assembly_report.txt` file. So, in those cases "AN:na" will already be what is written, and this just catches any cases where an entry might not be present in the `*assembly_report.txt`

Reply:
Update: CRAM building will fail when there are duplicate AN entries, so I'll need to add a check to remove these when present in `assembly_report.txt` and also remove writing the default when not present in lookup.

Reply:
I guess another implication of this is that we cannot use the `Assigned Molecule` field as a valid AN: identifier at all for any assembly that isn't 100% complete. I think it should be switched to use the `Sequence name` column instead, which also contains the molecule (e.g., SUPER_X), and is unique for each entry, although it requires parsing

Reply:
Oh gosh, I wasn't expecting that !
I wonder if there are rules such as "for all sequences with role=`assembled-molecule`, the "Assigned-molecule" name must be unique". Let's email Jo and Kerstin who are used to those reports and dealing with the NCBI. They can tell us more about it.

Reply:
Based on the email chain and looking at the `assembly_report.txt` for further examples I would suggest two options: 
1. Simply use the `Sequence name` field, which is unique 
2. If `role=assembled-molecule` name as `Assigned-molecule` (with `chr` prepended if this is generally expected -- this will work as well for cases where chromosomes have been chunked, as they would be named like `chr1_1` etc); otherwise use `Sequence name` 

I like option 2 better, although it might be more likely to break with assemblies not following the same naming conventions

Reply:
I like the option 2 better as well, but I'm not sure about adding the `chr` prefix. I can think of exceptions, like the mitochondrion being called `chrM` rather than `chrMT` (example: <https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_009663435.1/> (bottom of the chromosome list) vs <https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=calJac4&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrM%3A1%2D16499&hgsid=342945770_Ha2iNakxkcBAXgR0LjcyirTnaoHg> ). I'd rather stick with the default name that comes up in the assembly report.